velox/core/Expressions.h

```diff
@@ -327,6 +327,71 @@ class FieldAccessTypedExpr : public ITypedExpr {
 
 using FieldAccessTypedExprPtr = std::shared_ptr<const FieldAccessTypedExpr>;
 
+/// Represents a dereference expression which selects a subfield in a struct by
+/// index.
+class DereferenceTypedExpr : public ITypedExpr {
+ public:
+  DereferenceTypedExpr(TypePtr type, TypedExprPtr input, uint32_t index)
+      : ITypedExpr{std::move(type), {std::move(input)}}, index_(index) {
+    // Make sure this isn't being used to access a top level column.
+    VELOX_CHECK_NULL(
+        std::dynamic_pointer_cast<const InputTypedExpr>(inputs()[0]));
+  }
+
+  uint32_t index() const {
+    return index_;
+  }
+
+  const std::string& name() const {
+    return inputs()[0]->type()->asRow().nameOf(index_);
```

Comment:
Can we check this is not empty?  The result will most likely be ill-formed if we do want a name but get an empty string here (e.g. when we want to construct a `Subfield`)